File: src/0136.single-number/main.cpp
// Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be
// found in the LICENSE file.

#include <cassert>
#include <cstdio>

#include <vector>

class Solution {
 public:
  static int singleNumber(std::vector<int>& nums) {
    int current = 0;
    for (int num: nums) {
      current = current ^ num;
    }
    return current;
  }
};


void checkSolution() {
  {
    std::vector<int> nums = {2,2,1};
    assert(Solution::singleNumber(nums) == 1);
  }
  {
    std::vector<int> nums = {4,1,2,1,2};
    assert(Solution::singleNumber(nums) == 4);
  }
  {
    std::vector<int> nums = {1};
    assert(Solution::singleNumber(nums) == 1);
  }
}

int main() {
  checkSolution();
  return 0;
}

File: src/0136.single-number/src/main.rs
// Copyright (c) 2023 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be
// found in the LICENSE file.

// num ^ num = 0
pub fn single_number(nums: Vec<i32>) -> i32 {
    let mut ans = 0;
    for num in &nums {
        ans ^= num;
    }
    ans
}

pub fn single_number2(nums: Vec<i32>) -> i32 {
    const DIGIT_LEN: usize = 32;

    let mut ans = 0;
    for i in 0..DIGIT_LEN {
        let sum = nums.iter().map(|num| num >> i & 1).sum::<i32>();
        ans |= (sum % 2) << i;
    }
    ans
}

pub type SolutionFn = fn(Vec<i32>) -> i32;

fn check_solution(func: SolutionFn) {
    let nums = vec![2, 2, 1];
    assert_eq!(func(nums), 1);

    let nums = vec![4, 1, 2, 1, 2];
    assert_eq!(func(nums), 4);

    let nums = vec![1];
    assert_eq!(func(nums), 1);
}

fn main() {
    check_solution(single_number);
    check_solution(single_number2);
}

#[cfg(test)]
mod tests {
    use super::{check_solution, single_number, single_number2};

    #[test]
    fn test_single_number1() {
        check_solution(single_number);
    }

    #[test]
    fn test_single_number2() {
        check_solution(single_number2);
    }
}
